rasa/core/training/story_conflict.py

```diff
@@ -0,0 +1,221 @@
+        if last_event_type:
+            conflict_message = (
+                f"CONFLICT after {last_event_type} '{last_event_name}':\n"
+            )
+        else:
+            conflict_message = f"CONFLICT at the beginning of stories:\n"
+
+        # List which stories are in conflict with one another
+        for action, stories in self._conflicting_actions.items():
+            # Summarize if necessary
+            if len(stories) > 3:
+                # Four or more stories are present
+                conflict_description = (
+                    f"'{stories[0]}' and {len(stories) - 1} other trackers"
+                )
+            else:
+                conflict_description = (
+                    {1: "'{}'", 2: "'{}' and '{}'", 3: "'{}', '{}', and '{}'",}
+                    .get(len(stories))
+                    .format(*stories)
+                )
+
+            conflict_message += f"  {action} predicted in {conflict_description}\n"
+
+        return conflict_message
+
+
+TrackerEventStateTuple = namedtuple(
+    "TrackerEventStateTuple", "tracker event sliced_states"
+)
+
+
+def find_story_conflicts(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> List[StoryConflict]:
+    """Generates a list of StoryConflict objects, describing conflicts in the given trackers.
+
+    Args:
+        trackers: Trackers in which to search for conflicts.
+        domain: The domain.
+        max_history: The maximum history length to be taken into account.
+    Returns:
+        List of conflicts.
+    """
+
+    # We do this in two steps, to reduce memory consumption:
+
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_dict = _find_conflicting_states(trackers, domain, max_history)
+
+    # Iterate once more over all states and note the (unhashed) state,
+    # for which a conflict occurs
+    conflicts = _build_conflicts_from_states(
+        trackers, domain, max_history, state_action_dict
+    )
+
+    return conflicts
+
+
+def _find_conflicting_states(
+    trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+) -> Dict[int, Optional[List[Text]]]:
+    # Create a 'state -> list of actions' dict, where the state is
+    # represented by its hash
+    state_action_dict = defaultdict(list)
+    for element in _sliced_states_iterator(trackers, domain, max_history):
+        hashed_state = hash(str(list(element.sliced_states)))
+        if element.event.as_story_string() not in state_action_dict[hashed_state]:
+            state_action_dict[hashed_state] += [element.event.as_story_string()]
+
+    # Keep only conflicting `state_action_dict`s
+    return {
+        state_hash: actions
+        for (state_hash, actions) in state_action_dict.items()
+        if len(actions) > 1
+    }
+
+
+def _build_conflicts_from_states(
+    trackers: List["TrackerWithCachedStates"],
```

Comment:
it's already imported so the quotes don't add any advantage
```suggestion
    trackers: List[TrackerWithCachedStates],
```